File: dev/probe_writes.exs
# Probe: how long a write command takes on this server, and what a stall correlates with.
# Verdict: docs/decisions.md, "The write-stall verdict" — concurrent write tasks in the Docker
# Desktop VM; `--master local[1]` eliminates it. Re-run if the stall ever reappears.
#
#     docker compose up -d spark-connect
#     mix run dev/probe_writes.exs

defmodule ProbeWrites do
  def time(label, fun) do
    {us, result} = :timer.tc(fun)
    ms = us |> div(1000) |> Integer.to_string()
    status = if result == :ok, do: "ok", else: inspect(result, printable_limit: 120)

    IO.puts(
      String.pad_trailing(label, 44) <> String.pad_leading(ms <> " ms", 10) <> "  " <> status
    )

    result
  end
end

session = Latu.connect!("sc://localhost:15002")
IO.puts("Spark #{Latu.spark_version!(session)} — one write command per row, wall time\n")

df = Latu.range(session, 5)
base = "/tmp/latu_probe/#{System.unique_integer([:positive])}"

ProbeWrites.time("collect (baseline query, same session)", fn ->
  {:ok, _} = Latu.collect(df)
  :ok
end)

for n <- 1..3 do
  ProbeWrites.time("write parquet ##{n} to /tmp, same session", fn ->
    Latu.write(df, format: "parquet", path: "#{base}/p#{n}")
  end)
end

ProbeWrites.time("write csv to /tmp, same session", fn ->
  Latu.write(df, format: "csv", path: "#{base}/c1", header: true)
end)

session2 = Latu.connect!("sc://localhost:15002")
df2 = Latu.range(session2, 5)

for n <- 1..2 do
  ProbeWrites.time("write parquet ##{n}, FRESH session", fn ->
    Latu.write(df2, format: "parquet", path: "#{base}/s2_p#{n}")
  end)
end

ProbeWrites.time("save_as_table (warehouse bind mount)", fn ->
  Latu.save_as_table(df2, "latu_probe_#{System.unique_integer([:positive])}")
end)

Latu.disconnect(session)
Latu.disconnect(session2)

IO.puts("""

Reading the rows:
  only write #1 of each session stalls  ->  per-session setup (the artifact class loader)
  every write stalls                    ->  per-job, or the filesystem
  nothing stalls                        ->  needs the suite's churn; re-run during check.all
""")

File: dev/probe_writes.py
#!/usr/bin/env python3
"""Control for dev/probe_writes.exs: the same writes through PySpark.

The Elixir probe measured an intermittent ~60s lockstep stall in write stages (~1 run in 8).
PySpark stalling too means the server/container; PySpark never stalling across a few runs
means it is Latu's.

    export SPARK_REMOTE='sc://localhost:15002'
    dev/.venv/bin/python dev/probe_writes.py
"""

import os
import time

from pyspark.sql import SparkSession


def timed(label, fun):
    start = time.monotonic()
    fun()
    print(f"{label:<44}{(time.monotonic() - start) * 1000:>10.0f} ms")


def main():
    spark = SparkSession.builder.getOrCreate()
    print(f"Spark {spark.version} — one write per row, wall time\n")

    df = spark.range(5)
    base = f"/tmp/latu_probe_py/{int(time.time())}"

    timed("collect (baseline query, same session)", lambda: df.collect())

    # six samples: the stall hit roughly 1 in 8 writes on the Elixir side
    for n in range(1, 7):
        timed(f"write parquet #{n} to /tmp, same session",
              lambda n=n: df.write.parquet(f"{base}/p{n}"))

    timed("write csv to /tmp, same session",
          lambda: df.write.option("header", True).csv(f"{base}/c1"))

    spark2 = SparkSession.builder.remote(os.environ["SPARK_REMOTE"]).create()
    df2 = spark2.range(5)

    for n in (1, 2):
        timed(f"write parquet #{n}, FRESH session",
              lambda n=n: df2.write.parquet(f"{base}/s2_p{n}"))

    timed("save_as_table (warehouse bind mount)",
          lambda: df2.write.saveAsTable(f"latu_probe_py_{int(time.time())}"))


if __name__ == "__main__":
    main()
